src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+      }
+      case Coord.z: {
+        size.depth= value;
+        break;
+      }
+    }
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: origin,
+        size: size,
+        bytesPerRow: 1024,
+        rowsPerImage: 16,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+// TODO: might need dimensions.
+g.test('texture_range_conditions')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(poptions('origin_value', [7, 8]))
+      .combine(poptions('copy_size_value', [7, 8]))
+      .combine(poptions('texture_size_value', [14, 15]))
+      .combine(poptions('mipLevel', [0, 2]))
+      .combine(poptions('coordinate_to_test', kCoords))
+  )
+  .fn(async t => {
+    const { origin_value, copy_size_value, texture_size_value, mipLevel, coordinate_to_test, method } = t.params;
+
+    let origin = { x: 0, y: 0, z: 0 };
+    let copy_size = { width: 0, height: 0, depth: 0 };
+    let texture_size = { width: 16 << mipLevel, height: 16 << mipLevel, depth: 16 };
+    const success = origin_value + copy_size_value <= texture_size_value;
+
+    switch (coordinate_to_test) {
+      case Coord.x: {
+        origin.x = origin_value;
+        copy_size.width = copy_size_value;
+        texture_size.width = texture_size_value << mipLevel;
+        break;
+      }
+      case Coord.y: {
+        origin.y = origin_value;
+        copy_size.height = copy_size_value;
+        texture_size.height = texture_size_value << mipLevel;
+        break;
+      }
+      case Coord.z: {
+        origin.z = origin_value;
+        copy_size.depth = copy_size_value;
+        texture_size.depth = texture_size_value;
+        break;
+      }
+    }
+
+    console.log(origin.z, copy_size.depth, texture_size.depth);
+
+    const texture = t.device.createTexture({
+      size: texture_size,
+      mipLevelCount: 3,
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    t.expectValidationError(() => {
+      t.testRun({
+        texture: texture,
+        origin: origin,
+        size: copy_size,
+        mipLevel: mipLevel,
+        bytesPerRow: 1024,
+        rowsPerImage: 32,
+        dataSize: 1,
+        method: method,
+      });
+    }, !success);
+  });
+
+/*** Data-related validation: ***/
+
+g.test('texel_block_alignment_on_offset')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine([
+        { mul: 0, div: 1 }, // test 0
+        { mul: 1, div: 2 }, // test bytesPerBlock / 2
+        { mul: 1, div: 1 }, // test bytesPerBlock
+        { mul: 2, div: 1 }, // test 2 * bytesPerBlock
+        { mul: 3, div: 2 }, // test 3 * bytesPerBlock / 2
+        { mul: 3, div: 1 }, // test 3 * bytesPerBlock
```

Comment:
Perhaps just represent these in fixed-point, as multiples of "bytesPerBlock/2", like `poptions('halfBlocks', [0,1,2,3,4,6])`

Reply:
Done